File: core/src/main/java/com/spidey01/sxe/core/GameEngine.java
/*-
 * Copyright (c) 2012-current, Terry Mathew Poulin <BigBoss1964@gmail.com>
 *
 * This software is provided 'as-is', without any express or implied warranty.
 * In no event will the authors be held liable for any damages arising from the
 * use of this software.
 *
 * Permission is granted to anyone to use this software for any purpose,
 * including commercial applications, and to alter it and redistribute it
 * freely, subject to the following restrictions:
 *
 *	1. The origin of this software must not be misrepresented; you must
 *	   not claim that you wrote the original software. If you use this
 *	   software in a product, an acknowledgment in the product
 *	   documentation would be appreciated but is not required.
 *
 *	2. Altered source versions must be plainly marked as such, and must
 *	   not be misrepresented as being the original software.
 *
 *	3. This notice may not be removed or altered from any source
 *	   distribution.
 */

package com.spidey01.sxe.core;

import com.spidey01.sxe.core.common.Subsystem;
import com.spidey01.sxe.core.common.Utils;
import com.spidey01.sxe.core.config.Settings;
import com.spidey01.sxe.core.config.SettingsFile;
import com.spidey01.sxe.core.config.SettingsMap;
import com.spidey01.sxe.core.config.SettingsXMLFile;
import com.spidey01.sxe.core.graphics.Display;
import com.spidey01.sxe.core.input.InputManager;
import com.spidey01.sxe.core.logging.Log;
import com.spidey01.sxe.core.logging.Logging;
import com.spidey01.sxe.core.resource.ResourceManager;
import com.spidey01.sxe.core.sys.FileSystem;
import com.spidey01.sxe.core.sys.Platform;
import com.spidey01.sxe.core.sys.Xdg;

import java.io.BufferedReader;
import java.io.File;
import java.io.FileNotFoundException;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.io.PrintStream;
import java.util.ArrayList;
import java.util.Collection;
import java.util.Collections;
import java.util.HashMap;
import java.util.Iterator;
import java.util.List;
import java.util.Map;


/** Class implementing the core game engine.
 *
 *
 */
public class GameEngine {
    private static final String TAG = "GameEngine";

    private final Game mGame;
    private final Display mDisplay;
    private final SceneManager mSceneManager;
    private final InputManager mInputManager;
    private final ResourceManager mResourceManager;
    private final Logging mLogging;

    /** Master source of Settings.
     *
     * From lowest to highest priority, this will contain the following sources
     * of Settings data.
     *
     * <ol>
     *      <li>Platform settings.</li>
     *      <li>System settings file.</li>
     *      <li>User settings file.</li>
     *      <li>Command line arguments</li>
     * </ol>
     */
    private Settings mRuntimeSettings = new SettingsMap();

    /** Platform settings loaded from the ctor. */
    private Settings mPlatformSettings;

    /** System settings loaded from $XDG_CONFIG_DIRS. */
    private Settings mSystemSettings;

    /** User settings loaded from $XDG_CONFIG_HOME. */
    private Settings mUserSettings;

    /** Settings provided at the command line. */
    private final SettingsMap mCommandLineSettings;

    private final Platform mPlatform;
    private GameThread mGameThread;

    /** Master source of Subsystems registered with this GameEngine. */
    private List<Subsystem> mSubsystems = new ArrayList<Subsystem>();


    /** Initializes the engine for use.
     *
     * This is the primary chunk-o-code constructor.
     *
     * @param args Command line arguments.
     * @param display Display complicata.
     * @param scene  Manager of the scene.
     * @param game Game implementation.
     * @param input All that input related stuff.
     * @param resources Resource management.
     * @param settings Platform specific settings.
     * @param platform Platform specific information.
     */
    public GameEngine(SettingsMap args, Display display, SceneManager scene,
                      Game game, InputManager input, ResourceManager resources,
                      Logging logging, Settings settings, Platform platform)
    {
        mCommandLineSettings = args;
        mDisplay = display;
        mSceneManager = scene;
        mGame = game;
        mInputManager = input;
        mResourceManager = resources;
        mLogging = logging;
        mPlatformSettings = settings;
        mPlatform = platform;


        {
            assert mRuntimeSettings != null : "Programmer failure.";

            String cfgName = game.getName()+".cfg";
            String xmlName = game.getName()+".xml";
            String[] names = new String[]{ cfgName, xmlName };
            String p;
            File path;

            /*
             * System settings:
             *
             *      Look for the first GameName.cfg file in $XDG_CONFIG_DIRS.
             *      If not found, try GameName.xml by same method.
             */
            for (String n : names) {
                p = FileSystem.find(Xdg.XDG_CONFIG_DIRS, n);
                if (p == null) continue;
                path = new File(p);
                if (path.exists()) {
                    mSystemSettings = new SettingsFile(path);
                    break;
                }
            }

            /*
             * User settings:
             *
             *      Look for $XDG_CONFIG_HOME/|game name|cfg.
             *      If not found, try $XDG_CONFIG_HOME/|game name|.xml
             */
            for (String n : names) {
                path = new File(Xdg.XDG_CONFIG_HOME, n);
                if (path.exists()) {
                    mUserSettings = new SettingsFile(path);
                    break;
                }
            }

        }


        /*
         * These initialize functions will subscribe to whatever runtime
         * settings they want. As well as perform any pre start() setup.
         */
        mLogging.initialize(this);
        mDisplay.initialize(this);
        mSceneManager.initialize(this);
        mInputManager.initialize(this);
        mResourceManager.initialize(this);


        /*
         * Process the various sources of Settings.
         */
        if (mPlatformSettings != null)      mRuntimeSettings.merge(mPlatformSettings);
        if (mSystemSettings != null)        mRuntimeSettings.merge(mSystemSettings);
        if (mUserSettings != null)          mRuntimeSettings.merge(mUserSettings);
        if (mCommandLineSettings != null)   mRuntimeSettings.merge(mCommandLineSettings);

        configure();

        // ternary abuse, yeah.
        final String p = 
            (mGame == null ? "game" :
                (mPlatform == null ? "platform"
                    : null));

        if (p != null) {
            throw new IllegalArgumentException(p+" can't be null!");
        }
    }


    /** Start up the game.
     *
     * Takes care of initializing the games engine to run the game in the
     * current context. It will call the start() method of your Game
     * implementation accordingly.
     */
    public boolean start() {
        Log.v(TAG, "start()");

        Log.i(TAG, mPlatform);

        Log.i(TAG, "APP_HOME=\""+Xdg.APP_HOME+"\"");
        Log.i(TAG, "XDG_DATA_HOME=\""+Xdg.XDG_DATA_HOME+"\"");
        Log.i(TAG, "XDG_CONFIG_HOME=\""+Xdg.XDG_CONFIG_HOME+"\"");
        Log.i(TAG, "XDG_CACHE_HOME=\""+Xdg.XDG_CACHE_HOME+"\"");
        Log.i(TAG, "XDG_RUNTIME_DIR=\""+Xdg.XDG_RUNTIME_DIR+"\"");
        Log.i(TAG, "XDG_DATA_DIRS=\""+Utils.join(Xdg.XDG_DATA_DIRS, ':')+"\"");
        Log.i(TAG, "XDG_CONFIG_DIRS=\""+Utils.join(Xdg.XDG_CONFIG_DIRS, ':')+"\"");


        if (!mDisplay.create()) {
            return false;
        }

        mGameThread = new GameThread(this, mGame);
        mGameThread.start();

        return true;
    }


    /** Stop the game.
     *
     * Will ensure Game.stop() is called. Shuts down the display, etc.
     */
    public void stop() {
        mGame.stop();
        mGameThread.interrupt(); // should this be overriden to do Game.stop()?
		mDisplay.destroy();
        Log.v(TAG, "stop() done");
    }


    /** Convenience method that can serve as a simple main loop. */
    public void mainLoop() {
        try {
            while (isRunning()) {
                update();
            }
        } catch(Exception e) {
            Log.e(TAG, "Unhandled exception in mainLoop().", e);
            // Don't call stop. We don't know if it's that recoverable.
            // Dump a stack trace and force an exit NOW.
            e.printStackTrace();
            System.exit(1);
        }
    }


    /** Determine if the Game environment is running. */
    public boolean isRunning() {
        return (!mGame.isStopRequested()
                && !mDisplay.isCloseRequested()
                && mGameThread.isAlive());
    }


    /** Update Game engine state.
     *
     * <ol>
     *  <li>Polls the InputManager.</li>
     *  <li>Updates the SceneManager.</li>
     *  <li>Updates the Display.</li>
     * </ol>
     *
     * Game implementations update independantly of GameEngine, as it runs on a
     * seperate thread.
     */
    public void update() {
            mInputManager.poll();
            mSceneManager.update();
            mDisplay.update();
    }


    public Display getDisplay() {
        return mDisplay;
    }

    
    public SceneManager getSceneManager() {
        return mSceneManager;
    }


    public Game getGame() {
        return mGame;
    }


    public InputManager getInputManager() {
        return mInputManager;
    }


    public ResourceManager getResourceManager() {
        return mResourceManager;
    }


    public Settings getSettings() {
        return mRuntimeSettings;
    }


    public void addSubsystem(Subsystem subsystem) {
        mSubsystems.add(subsystem);
    }


    public void addSubsystems(Collection<? extends Subsystem> subsystems) {
        mSubsystems.addAll(subsystems);
    }


    public void removeSubsystem(Subsystem subsystem) {
        mSubsystems.add(subsystem);
    }


    public void removeSubsystem(String name) {
        for (Iterator<?> it = mSubsystems.iterator(); it.hasNext(); ) {
            if (name.equals(it.next())) {
                it.remove();
                /*
                 * Don't think you can optimize away the rest of the loop.
                 * Nothing says that two subsystem cannot have the same name!
                 */
            }
        }
    }


    public Subsystem getSubsystem(Subsystem subsystem) {
        return mSubsystems.get(mSubsystems.indexOf(subsystem));
    }


    /** Get registered Subsystem by name.
     *
     * @param name expected value of {@link Subsystem#name()}.
     * @return First subsystem matching name or null if not found.
     */
    public Subsystem getSubsystem(String name) {
        for (Subsystem subsystem : mSubsystems) {
            if (name.equals(subsystem.name())) {
                return subsystem;
            }
        }
        return null;
    }


    private Subsystem getSubsystem(int index) {
        return mSubsystems.get(index);
    }


    private List<Subsystem> getSubysystems() {
        return Collections.unmodifiableList(mSubsystems);
    }


    public void configure() {
        configure(mRuntimeSettings);
    }


    public void configure(Settings s) {
        /* Shortcut key used for general debugging. */
        if (s.getBoolean("debug")) {
            // Make sure that we have a log file.
            if (!s.contains("debug.log_to"))
                s.setString("debug.log_to", "debug.log");
            if (!s.contains("debug.log_level"))
                s.setInt("debug.log_level", Log.DEBUG);
        }
    }


}

File: core/src/GameEngine.cpp
/*-
 * Copyright (c) 2012-current, Terry Mathew Poulin <BigBoss1964@gmail.com>
 *
 * This software is provided 'as-is', without any express or implied warranty.
 * In no event will the authors be held liable for any damages arising from the
 * use of this software.
 *
 * Permission is granted to anyone to use this software for any purpose,
 * including commercial applications, and to alter it and redistribute it
 * freely, subject to the following restrictions:
 *
 *	1. The origin of this software must not be misrepresented; you must
 *	   not claim that you wrote the original software. If you use this
 *	   software in a product, an acknowledgment in the product
 *	   documentation would be appreciated but is not required.
 *
 *	2. Altered source versions must be plainly marked as such, and must
 *	   not be misrepresented as being the original software.
 *
 *	3. This notice may not be removed or altered from any source
 *	   distribution.
 */

#include "sxe/core/GameEngine.hpp"

#include <sxe/core/Game.hpp>
#include <sxe/core/config/Settings.hpp>
#include <sxe/core/config/SettingsMap.hpp>
#include <sxe/core/config/SettingsXMLFile.hpp>
#include <sxe/core/graphics/Display.hpp>
#include <sxe/core/sys/Platform.hpp>
#include <sxe/logging.hpp>

using std::string;

namespace sxe { namespace core {

const string GameEngine::TAG = "GameEngine";

GameEngine::GameEngine()
{
#if 0
    // placeholder
    auto sink = std::make_shared<sxe::core::logging::LogSink>();
    sxe::core::logging::Log::add(sink);
#endif
}

GameEngine::GameEngine(Game_ptr game, Settings_ptr&& args,
                       DisplayManager_ptr&& display, SceneManager_ptr&& scene,
                       InputManager_ptr&& input, ResourceManager_ptr&& resources,
                       LoggingManager_ptr&& logging, Settings_ptr&& settings,
                       sys::Platform platform)
    : mXdg()
    , mGame(game)
    , mDisplayManager(std::move(display))
    , mSceneManager(std::move(scene))
    , mInputManager(std::move(input))
    , mResourceManager(std::move(resources))
    , mLoggingManager(std::move(logging))
    , mRuntimeSettings(std::make_unique<config::SettingsMap>())
    , mPlatformSettings(std::move(settings))
    , mSystemSettings(nullptr)
    , mUserSettings(nullptr)
    , mCommandLineSettings(std::move(args))
    , mPlatform(platform)
{

    if (mGame == nullptr) {
        throw std::invalid_argument(TAG + string("game parameter can't be nullptr!"));
    }

    // placeholder
    auto sink = std::make_shared<sxe::core::logging::LogSink>();
    sxe::core::logging::Log::add(sink);

    {
        if (!mRuntimeSettings)
            Log::wtf(TAG, "Programmer failure");

        // Used to have a fancier SettingsFile stuff for k/v vs xml - start off with what we have now.

        const string cfgName = mGame->getName() + ".cfg";
        const string xmlName = mGame->getName() + ".xml";
        const auto names = { cfgName, xmlName };

        /*
         * System settings:
         *
         *      Look for the first GameName.cfg file in $XDG_CONFIG_DIRS.
         *      If not found, try GameName.xml by same method.
         */

        for (const string& n : names) {
            auto p = mXdg.getConfigDir(n);

            if (sys::FileSystem::exists(p)) {
                mSystemSettings = std::make_unique<config::SettingsXMLFile>(p);
                break;
            }
        }

        /*
         * User settings:
         *
         *      Look for $XDG_CONFIG_HOME/|game name|cfg.
         *      If not found, try $XDG_CONFIG_HOME/|game name|.xml
         */

        for (const string& n : names) {
            auto p = mXdg.getConfigHomeDir(n);

            if (sys::FileSystem::exists(p)) {
                mUserSettings = std::make_unique<config::SettingsXMLFile>(p);
            }
        }
    }

    /*
     * These initialize functions will subscribe to whatever runtime
     * settings they want. As well as perform any pre start() setup.
     */

    if (mLoggingManager)
        mLoggingManager->initialize(*this);
    if (mDisplayManager)
        mDisplayManager->initialize(*this);
    if (mSceneManager)
        mSceneManager->initialize(*this);
    if (mInputManager)
        mInputManager->initialize(*this);
    if (mResourceManager)
        mResourceManager->initialize(*this);

    /*
     * Process the various sources of Settings.
     */
    if (mPlatformSettings != nullptr)      mRuntimeSettings->merge(*mPlatformSettings);
    if (mSystemSettings != nullptr)        mRuntimeSettings->merge(*mSystemSettings);
    if (mUserSettings != nullptr)          mRuntimeSettings->merge(*mUserSettings);
    if (mCommandLineSettings != nullptr)   mRuntimeSettings->merge(*mCommandLineSettings);

    configure();
}


GameEngine::~GameEngine()
{
    if (mResourceManager)
        mResourceManager->uninitialize();
    if (mInputManager)
        mInputManager->uninitialize();
    if (mSceneManager)
        mSceneManager->uninitialize();
    if (mDisplayManager)
        mDisplayManager->uninitialize();
    if (mLoggingManager)
        mLoggingManager->uninitialize();
}


bool GameEngine::start()
{
    Log::v(TAG, "start()");

    // XXX don't have platform settings yet.
    // Log.i(TAG, mPlatform);

    Log::i(TAG, "APP_HOME=\"" + mXdg.APP_HOME.string() + "\"");
    Log::i(TAG, "XDG_DATA_HOME=\"" + mXdg.XDG_DATA_HOME.string() + "\"");
    Log::i(TAG, "XDG_CONFIG_HOME=\"" + mXdg.XDG_CONFIG_HOME.string() + "\"");
    Log::i(TAG, "XDG_CACHE_HOME=\"" + mXdg.XDG_CACHE_HOME.string() + "\"");
    Log::i(TAG, "XDG_RUNTIME_DIR=\"" + mXdg.XDG_RUNTIME_DIR.string() + "\"");
    // Log.i(TAG, "XDG_DATA_DIRS=\""+Utils.join(Xdg.XDG_DATA_DIRS, ':')+"\"");
    // Log.i(TAG, "XDG_CONFIG_DIRS=\""+Utils.join(Xdg.XDG_CONFIG_DIRS, ':')+"\"");

    if (!mDisplayManager) {
        Log::e(TAG, "No Display implementation!");
        Log::w(TAG, "If your platform or app has none: use NullDisplay.");
        return false;
    }

    if (!mDisplayManager->create()) {
        Log::e(TAG, "Display::create() failed!");
        return false;
    }

    #if 0 // 1.x did:
    mGameThread = new GameThread(this, mGame);
    mGameThread.start();
    #endif

    return true;
}


void GameEngine::stop()
{
    #if 0 // 1.x
    mGame.stop();
    mGameThread.interrupt(); // should this be overriden to do Game.stop()?
    #endif
    mDisplayManager->destroy();
    Log::v(TAG, "stop() done");
}


void GameEngine::mainLoop()
{
    try {
        while (isRunning()) {
            update();
        }
    } catch(std::exception& ex) {
        Log::e(TAG, "Unhandled exception in mainLoop().", ex);

        /*
         * Don't call stop. We don't know if it's that recoverable. Dump a
         * stack trace and force an exit NOW.
         */
        #if 0 // 1.x
        e.printStackTrace();
        #endif
        std::exit(1);
    }
}


bool GameEngine::isRunning() const
{
    if (mGame && mGame->isStopRequested())
        return false;

    if (mDisplayManager && mDisplayManager->isCloseRequested())
        return false;

    return true;
    #if 0 // 1.x
        return (!mGame.isStopRequested()
                && !mDisplay.isCloseRequested()
                && mGameThread.isAlive());
    #endif
}


void GameEngine::update()
{
    #if 0 // 1.x
    mInputManager.poll();
    mSceneManager.update();
    #endif
    if (mDisplayManager)
        mDisplayManager->update();
}


void GameEngine::configure()
{
    configure(*mRuntimeSettings);
}


void GameEngine::configure(config::Settings& s)
{
    /* Shortcut key used for general debugging. */
    if (s.getBool("debug")) {
        // Make sure that we have a log file.
        if (!s.contains("debug.log_to"))
            s.setString("debug.log_to", "debug.log");
        if (!s.contains("debug.log_level"))
            s.setInt("debug.log_level", Log::DEBUG);
    }
}


std::weak_ptr<Game> GameEngine::getGame() const
{
    return mGame;
}



config::Settings& GameEngine::getSettings() const
{
    assert(mRuntimeSettings != nullptr);

    return *mRuntimeSettings;
}


} }

